## what is the function ?
## User defined functions
- Afunction is a reusable block of code that performs a specifice tasks . 
- Instead of writing the same code multiple times , you write it onces in function and call it whenever needed.
-
-  Less code duplication 
- Easier to maintain 
- more reables 
- Reuable across projects 

In [2]:
## calculate avg of three diff groups 

group1_sum = 12 + 34 + 67 
group1_avg = group1_sum / 3  
print(f"Group 1 average: {group1_avg}")


group2_sum = 45 + 67 + 23
group2_avg = group2_sum / 3
print(f"Group 2 average: {group2_avg}")

group3_sum = 43 + 78 + 89
group3_avg = group3_sum / 3
print(f"Group 3 average: {group3_avg}")

Group 1 average: 37.666666666666664
Group 2 average: 45.0
Group 3 average: 70.0


In [3]:
# with functions 
def calculate_average(scores):
    total = sum(scores)
    avg = total / len(scores)
    return avg
print(calculate_average([1,2,3]))

2.0


In [4]:
calculate_average([566,90,8766])

3140.6666666666665

In [ ]:
## Basic syntax for defining fucntiosn 
def function_name(parameters):
    # code block
    # statements
    return value

In [1]:
def add(a,b):
    result = a + b 
    return result 

In [2]:
# calling a function
add (5,4)

9

In [3]:
ans = add(5,4)
print(ans)

9


In [ ]:
# Function Without Parameters
def greet():
    print("i love coding")

In [ ]:
greet()

i love coding


In [ ]:
greet()

i love coding


In [ ]:
help(greet)

Help on function great in module __main__:

great()
    # Function Without Parameters



In [8]:
# Function Without Return
def say_hi(name):
    print("Hi", name)
say_hi("Swarn")


Hi Swarn


In [9]:
def msg(name , great):
    print("hello ",name , great)
msg("Swarna ", "Good Moring")    

hello  Swarna  Good Moring


In [10]:
msg("Bhanu", "Good Night")

hello  Bhanu Good Night


In [11]:
def factorial(n):
    if n == 1:
        return 1
    return n * factorial(n-1)

print(factorial(5))

120


## Write a function to add two numbers.
- Input: 5, 10
- Output: 15

In [12]:
def add(a,b):
    c = a+ b 
    return c

In [13]:
add (5,10)

15

### Write a function to find the square of a number.
- Input: 4
- Output: 16

In [16]:
def squ(n):
    x = n**2
    return x

In [17]:
squ(4)

16

## Write a function that prints "Hello, Python!".


In [18]:
def fun(name):
    print("hello",name)
    return 

In [19]:
fun("python")

hello python


## Write a function to check if a number is even or odd.
- Input: 7
- Output: Odd

In [20]:
def check(num):
    if num %2==0:
        print("Even")
    else:
        print("Odd")  
    return          

In [21]:
check(7)

Odd


## Write a function to find the factorial of a number.
- Input: 5
- Output: 120

In [22]:
def factorial(n):
    if n==0:
        return 1
    return n* factorial(n-1)

In [23]:
factorial(5)

120

In [25]:
# Write a function to find the maximum of two numbers
def maximum(a, b):
    if a > b:
        return a
    else:
        return b

In [26]:
maximum(23,43)

43

In [27]:
# Write a function to calculate the area of a rectangle.
def area_rectangle(length,width):
    area = length * width
    return area

In [28]:
area_rectangle(2,3)

6

In [34]:
def count_vowels(s):
    count = 0
    for ch in s:
        if ch.lower() in "aeiou":
            count += 1
    return count

In [35]:
count_vowels("Swarma")

2

In [41]:
# Reverse a number
def rev(n):
    reverse = 0
    while n > 0:
        digit = n % 10
        reverse = reverse * 10 + digit
        n //= 10
    return reverse

In [43]:
rev(1234)

4321

## Key Points

def keyword is used to define a function

Function name should be meaningful

Indentation is mandatory

return sends a value back to the caller

A function can be called multiple times

## Parameter and arguments 

- Parameters = Variables listed in the fucntion definition 
- Arguments = Actual values passed when calling the fucntion 